v1/plugins/logs/encoder.go

```diff
@@ -71,165 +91,376 @@ func (enc *chunkEncoder) WithMetrics(m metrics.Metrics) *chunkEncoder {
+					len(eventBytes), enc.limit)
+			}
+			return nil, nil
 		}
 
-		result, err = enc.reset()
+		// re-encode the event with the ND cache removed
+		event.NDBuiltinCache = nil
+		if enc.logger != nil {
+			enc.logger.Error("ND builtins cache dropped from this event to fit under maximum upload size limits. Increase upload size limit or change usage of non-deterministic builtins.")
+		}
+		enc.incrMetric(logNDBDropCounterName)
+
+		var err error
+		eventBytes, err = json.Marshal(&event)
 		if err != nil {
 			return nil, err
 		}
 	}
 
+	if int64(len(eventBytes)+enc.bytesWritten+1) <= enc.uncompressedLimit {
+		return nil, enc.appendEvent(eventBytes)
+	}
+
+	// Adjust the encoder's uncompressed limit based on the current amount of
+	// data written to the underlying buffer. The uncompressed limit decides when to return a chunk.
+	// The uncompressed limit is modified based on the below algorithm:
+	// 1) Scale Up: If the current chunk size is within 90% of the user-configured limit, exponentially increase
+	// the uncompressed limit. The exponential function is 2^x where x has a minimum value of 1.
+	// A chunk will be returned with what was already written, the buffer was underutilized but the next time it shouldn't be.
+	// The incoming event is written to the next chunk.
+	// 2) Scale Down: If the current chunk size exceeds the compressed limit, decrease the uncompressed limit and re-encode the
+	// decisions in the last chunk.
+	// 3) Equilibrium: If the chunk size is between 90% and 100% of the user-configured limit, maintain uncompressed limit value.
+	// A chunk will be returned with what was already written, the uncompressed limit is ideal.
+	// The incoming event is written to the next chunk.
+
+	// The uncompressed size is too small (it starts equal to the compressed limit)
+	// Or this is a recursive call to Write trying to split the events into separate chunks
 	if enc.bytesWritten == 0 {
-		n, err := enc.w.Write([]byte(`[`))
+		// If an event is too large, there are multiple things to try before dropping the event:
+		// 1. Try to fit the incoming event into the next chunk without losing ND cache
+		if err := enc.appendEvent(eventBytes); err != nil {
+			return nil, err
+		}
+
+		result, err := enc.reset()
 		if err != nil {
 			return nil, err
 		}
-		enc.bytesWritten += n
-	} else {
-		n, err := enc.w.Write([]byte(`,`))
+
+		currentSize := len(result[0])
+		if currentSize < int(enc.limit) {
+			// success! the incoming chunk doesn't have to lose the ND cache and can go into a chunk by itself
+			// scale up the uncompressed limit using the uncompressed event size as a base
+			err = enc.appendEvent(eventBytes)
+			if err != nil {
+				return nil, err
+			}
+			enc.uncompressedLimit = int64(len(eventBytes))
+			enc.scaleUp()
+			return nil, nil
+		}
+
+		// The ND cache has to be dropped, record this size as a known maximum event size
+		if enc.maxEventSize == 0 || int64(len(eventBytes)) < enc.maxEventSize {
```

Comment:
So `maxEventSize` is a misnomer? This isn't the maximum allowed size, this is some point beyond the max?

Reply:
Who knows 😄. Let's keep it, and revise in the future if needed.